### DISCUSSION TITLE
Transparent shadow quality

### DISCUSSION THREAD

**Discussion by wrohv** [google/filament#5048] [Q&A] [answered]
In v1.16.0, I am seeing a moire pattern in transparent shadows. For example, the one on the left here:
![AxesPCF](https://user-images.githubusercontent.com/43356887/149414800-cd016a0e-85f7-4923-ac56-fed84637c77c.jpg)

This didn't seem to be as much of an issue before I upgraded from v1.12.9:
![AxesPCF_1129](https://user-images.githubusercontent.com/43356887/149414957-32acc17e-1c0b-45f7-a5ab-d60bf940f6f2.jpg)

Here I have 3 spot lights, PCF shadows, macOS+arm; identical behavior on OpenGL and Metal.

I've tried a lot of different settings; haven't yet found anything to significantly improve it. Suggestions?

**Comment by romainguy** [comment]
1.16.0 doesn't use PCF anymore, only VSM. It's probably because of that. @pixelflinger, we probably need to rethink how we dither transparent shadows with VSM.

**Reply by pixelflinger** [reply]
No we are still using PCF be default, but it's only "hard shadows".

Actually a way to completely fix this is to use VSM + small blur.

**Reply by romainguy** [reply]
Ah then I'm not sure why the dither would look different.

**Reply by wrohv** [reply]
Just switching to VSM doesn't change much, but adding `vsm.blurWidth=10` improves it, but the blurring is significant. The moire artifact is still clearly visible at a blurWidth of 2.

Blur 10:
![AxesVSMBlur10](https://user-images.githubusercontent.com/43356887/149419101-47fde5f5-1de8-49c1-a0dd-48bb86c8b0c8.jpg)

Blur 2:
![AxesVSMBlur2](https://user-images.githubusercontent.com/43356887/149419410-b36e8e1e-8ce3-4cf1-b8a6-af53af37263a.jpg)

So I'm not yet seeing an option in 1.16.0 that works as well as the prior version.

**Comment by pixelflinger** [comment, marked as answer]
VSM + blur filter of size 3 should be equivalent to the old PCF algorithm.
However, when I tried this I still see a lot of moiré, I think it is because of how VSM works. In my experience you need at least a blur of 9 to get rid of the artifacts.

I might bring back the old PCF code for this and performance reasons. stay tunned.

**Reply by wrohv** [reply]
I also see a blur of ~9 being needed to remove the moire. Thanks for your help!

**Reply by romainguy** [reply]
Yes that matches the size of the dithering pattern.

**Reply by pixelflinger** [reply]
Actually the dithering pattern is a gradient noise, so it doesn't really have a "pattern" or size. But strangely, 9 seems to work.

**Reply by wrohv** [reply]
Just confirming that the previous behavior for PCF shadows is restored in v1.17.0. Thanks!
